docs/documentation/code-format-lint.md

```diff
@@ -0,0 +1,98 @@
+# Linting and Code Formatting
+
+In our [packaging guide](https://www.pyopensci.org/python-package-guide/package-structure-code/code-style-linting-format.html), we
+discuss various linters and code formatters used in the scientific Python
+ecosystem, including tools like:
+
+* [black](https://github.com/psf/black)
+* [flake8](https://github.com/pycqa/flake8)
+* [isort](https://pycqa.github.io/isort/)
+
+[Ruff](https://github.com/charliermarsh/ruff) is a tool that has been
+quickly becoming more popular. It is written in the Rust programming language
+and is fast. We like Ruff because it can perform both code formatting
+(similar to Black) and linting (similar to Flake8). This makes your setup
+simpler.
+
+## Running Linters with pre-commit
+
+There are a few ways to call the linters when you are working on your code.
+
+One option is to use pre-commit hooks. [pre-commit](https://pre-commit.com/)
+runs any defined linters, code and text formatters, spellcheckers, and other
+tools on your code locally when you use `git commit` to make a change. For
+example:
+
+```bash
+git commit -m "message here"
+```
+
+By configuring pre-commit hooks, you can automatically run tools like Ruff and
+codespell on your code and documentation every time you commit. This ensures
+consistency and catches errors early.
+
+While pre-commit is a powerful tool to add to your workflow, many do not like it:
+
+* For beginner contributors, running pre-commit hooks can be confusing. You need
+  to understand that each time you commit, it will run the checks. If it finds
+  issues in your files, it will not actually commit your files to history. This
+  can be confusing for even seasoned developers if they haven't used pre-commit
+  before.
+* Some prefer to set up autoformatters that run every time you save a file in
```

Comment:
We've mitigated this in some projects by relying on standard config as much as possible so that the scripts/IDEs can be run independently of `pre-commit`, and leaning on `hatch` environments for complex tasks.

Reply:
so helpful - i think the same question as above - could you perhaps direct me to an example of how you run linters / formatter? i could see a hatch lint that we create or... ?

Reply:
Thank you for this @blink1073 !!